# Calories Restriction & Weight Loss

## Imports

In [8]:
import os

import pandas as pd
import numpy as np
from scipy import stats
import statsmodels.api as sm

## Data

### Load

In [9]:
data_path = './data'
data_files = os.listdir(data_path)
data = pd.DataFrame(columns=['burned', 'steps', 'weight', 'consumed'])
dates = None
for fname in data_files:
    if 'activity' in fname:
        activity_data = pd.read_csv(os.path.join(data_path, fname))
        data.loc[:, 'burned'] = activity_data['Calories Burned']
        data.loc[:, 'steps'] = activity_data['Steps']
        dates = activity_data['Date']
    elif 'body' in fname:
        data_body = pd.read_csv(os.path.join(data_path, fname))
        data.loc[:, 'weight'] = data_body['Weight']
    elif 'food' in fname:
        data_food = pd.read_csv(os.path.join(data_path, fname))
        data.loc[:, 'consumed'] = data_food['Calories In']
data.index = dates
data['deficite'] = data['burned'] - data['consumed']
data['weight_delta'] = data['weight'].diff(periods=1).shift(-1)
data = data.dropna()

In [10]:
data

burned  steps  weight  consumed  deficite  weight_delta
Date                                                               
2019-01-01    2873   7065   67.70         0      2873          0.00
2019-01-02    2163   3715   67.70      1561       602          0.00
2019-01-03    2670   8142   67.70      1981       689          0.00
2019-01-04    2500   7549   67.70      2090       410          0.00
2019-01-05    2885  14040   67.70      3112      -227          0.00
2019-01-06    2044   2765   67.70       378      1666          0.00
2019-01-07    2594   7827   67.70      2010       584          1.15
2019-01-08    2270   7715   68.85      1916       354          1.15
2019-01-09    3016  12465   70.00      3254      -238          0.00
2019-01-10    2812  10632   70.00      2130       682         -0.20
2019-01-11    2966  11243   69.80      2419       547          0.40
2019-01-12    2700   7714   70.20      2378       322         -0.12
2019-01-13    2827  10108   70.08      2668       159         -0.12
2019-01-14    2104   3952   69.96      2102         2         -0.12
2019-01-15    2377   8035   69.84      1921       456         -0.12
2019-01-16    2320   4194   69.72      1593       727         -0.12
2019-01-17    2355   5668   69.60      1765       590         -0.60
2019-01-18    3157  15516   69.00      1411      1746          0.70
2019-01-19    2540   2532   69.70         0      2540          0.70
2019-01-20    2419   6342   70.40         0      2419          0.70
2019-01-21    2413   8494   71.10      1246      1167         -1.50
2019-01-22    2661   6436   69.60      1922       739         -0.20
2019-01-23    2685  11564   69.40      1522      1163         -0.70
2019-01-24    2420   5212   68.70      1942       478         -0.30
2019-01-25    2598   9732   68.40      1871       727         -0.04
2019-01-26    2468   6856   68.36      2590      -122          0.54
2019-01-27    2071   3187   68.90      2180      -109         -0.20
2019-01-28    2378   4899   68.70      1457       921         -0.30
2019-01-29    2436   8510   68.40      1759       677         -0.04
2019-01-30    2011   2444   68.36      1229       782         -0.66
2019-01-31    2887  11569   67.70      1989       898         -0.02
2019-02-01    2489   9430   67.68      1717       772          0.22
2019-02-02    2042   2437   67.90      2267      -225          0.30
2019-02-03    2290   5939   68.20      1954       336          0.60
2019-02-04    2342   7581   68.80      1202      1140         -1.20
2019-02-05    2728   8823   67.60      2345       383         -0.10
2019-02-06    1922   1880   67.50      1182       740         -0.20
2019-02-07    2230   2816   67.30      1429       801         -0.20
2019-02-08    2450   8523   67.10      2288       162         -0.20
2019-02-09    2781   9412   66.90      3139      -358          1.80
2019-02-10    2212   4219   68.70      1250       962         -0.90
2019-02-11    2622   5395   67.80      1605      1017          0.60
2019-02-12    3021  11369   68.40      1719      1302         -0.70
2019-02-13    2153   4030   67.70      1384       769         -0.80
2019-02-14    2849  14274   66.90      2830        19          1.00
2019-02-15    2333   8587   67.90      1595       738         -0.80
2019-02-16    2303   6481   67.10      1702       601          0.20
2019-02-17    2429   8655   67.30      1648       781         -0.40
2019-02-18    2421   8111   66.90      1410      1011         -0.50
2019-02-19    2473   7435   66.40      1613       860         -0.20
2019-02-20    2683   7063   66.20      2019       664          0.40
2019-02-21    2540   8852   66.60      1557       983         -0.70
2019-02-22    2542  12118   65.90      1823       719         -0.10
2019-02-23    2197   5390   65.80      1846       351          0.20
2019-02-24    1902   1948   66.00      1973       -71          0.20
2019-02-25    1937   2037   66.20      1358       579         -0.45
2019-02-26    2365   7956   65.75      1515       850         -0.45


### Clean

In [11]:
data = data.loc[~(data == 0).any(1)]  # remove rows containing zero value
stats = data.describe().loc[['mean', 'std']]
outliers = (np.abs(stats.loc['mean'] - data) > stats.loc['std'] * 2).any(1)
data = data.loc[~outliers]

In [12]:
data

burned  steps  weight  consumed  deficite  weight_delta
Date                                                               
2019-01-07    2594   7827   67.70      2010       584          1.15
2019-01-08    2270   7715   68.85      1916       354          1.15
2019-01-10    2812  10632   70.00      2130       682         -0.20
2019-01-11    2966  11243   69.80      2419       547          0.40
2019-01-12    2700   7714   70.20      2378       322         -0.12
2019-01-13    2827  10108   70.08      2668       159         -0.12
2019-01-14    2104   3952   69.96      2102         2         -0.12
2019-01-15    2377   8035   69.84      1921       456         -0.12
2019-01-16    2320   4194   69.72      1593       727         -0.12
2019-01-17    2355   5668   69.60      1765       590         -0.60
2019-01-22    2661   6436   69.60      1922       739         -0.20
2019-01-23    2685  11564   69.40      1522      1163         -0.70
2019-01-24    2420   5212   68.70      1942       478         -0.30
2019-01-25    2598   9732   68.40      1871       727         -0.04
2019-01-26    2468   6856   68.36      2590      -122          0.54
2019-01-27    2071   3187   68.90      2180      -109         -0.20
2019-01-28    2378   4899   68.70      1457       921         -0.30
2019-01-29    2436   8510   68.40      1759       677         -0.04
2019-01-30    2011   2444   68.36      1229       782         -0.66
2019-01-31    2887  11569   67.70      1989       898         -0.02
2019-02-01    2489   9430   67.68      1717       772          0.22
2019-02-02    2042   2437   67.90      2267      -225          0.30
2019-02-03    2290   5939   68.20      1954       336          0.60
2019-02-04    2342   7581   68.80      1202      1140         -1.20
2019-02-05    2728   8823   67.60      2345       383         -0.10
2019-02-06    1922   1880   67.50      1182       740         -0.20
2019-02-07    2230   2816   67.30      1429       801         -0.20
2019-02-08    2450   8523   67.10      2288       162         -0.20
2019-02-10    2212   4219   68.70      1250       962         -0.90
2019-02-11    2622   5395   67.80      1605      1017          0.60
2019-02-12    3021  11369   68.40      1719      1302         -0.70
2019-02-13    2153   4030   67.70      1384       769         -0.80
2019-02-15    2333   8587   67.90      1595       738         -0.80
2019-02-16    2303   6481   67.10      1702       601          0.20
2019-02-17    2429   8655   67.30      1648       781         -0.40
2019-02-18    2421   8111   66.90      1410      1011         -0.50
2019-02-19    2473   7435   66.40      1613       860         -0.20
2019-02-20    2683   7063   66.20      2019       664          0.40
2019-02-21    2540   8852   66.60      1557       983         -0.70
2019-02-22    2542  12118   65.90      1823       719         -0.10
2019-02-23    2197   5390   65.80      1846       351          0.20
2019-02-24    1902   1948   66.00      1973       -71          0.20
2019-02-25    1937   2037   66.20      1358       579         -0.45
2019-02-26    2365   7956   65.75      1515       850         -0.45

## Model

In [14]:
X = data['deficite'].values.reshape(-1, 1)
y = data['weight_delta'].values
X_const = sm.add_constant(X)
model = sm.OLS(y, X_const)
res = model.fit()

In [15]:
loss_500 = res.params[1] * 500
loss_500

-0.326366203781185

In [16]:
conf_int = loss_500 + (res.conf_int(.05)[1] - res.params[1])
conf_int

array([-0.32674504, -0.32598737])

In [17]:
res.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.224
Model:                            OLS   Adj. R-squared:                  0.205
Method:                 Least Squares   F-statistic:                     12.09
Date:                Thu, 02 May 2019   Prob (F-statistic):            0.00119
Time:                        12:13:14   Log-Likelihood:                -25.630
No. Observations:                  44   AIC:                             55.26
Df Residuals:                      42   BIC:                             58.83
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.2658      0.132      2.007      0.051      -0.002       0.533
x1            -0.0007      0.000     -3.477      0.001      -0.001      -0.000
==============================================================================
Omnibus:                        9.952   Durbin-Watson:                   1.687
Prob(Omnibus):                  0.007   Jarque-Bera (JB):                9.219
Skew:                           0.997   Prob(JB):                      0.00996
Kurtosis:                       4.027   Cond. No.                     1.40e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.4e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

The weight_delta vairiable has a statistically-significant effect within the model.

The model predicts a loss of 326 grams (-0.326, 95% CI [-0.327, -.326]) for a day of 500 calories deficit.